content/en/blog/2025/contrib-unroll-processor.md

```diff
@@ -0,0 +1,165 @@
+
+The unroll processor cleanly separates the concern of adding records from
+transformation logic and operates in a way that's both composable and
+predictable.
+
+I helped develop the first version of the unroll processor in the Bindplane
+Distro of OpenTelemetry Collector. It was first shipped and in use by customers
+in January 2025 and has been running in production ever since.
+
+I've seen customers use it across:
+
+- VPC logs
+- CloudWatch ingestion pipelines
+- Windows + endpoint logs
+- Bundled collector telemetry
+
+We observed very low issue volume even under real production load, which gave us
+the confidence to propose it upstream. Specifically when the initial receiver or
+source of the log signals is fairly format agnostic.
+
+## How to configure the unroll processor
+
+Drop the unroll processor into your pipeline wherever you need to expand bundled
+log payloads. Here's a minimal configuration example to get you started:
+
+```yaml
+processors:
+  unroll:
+service:
+  pipelines:
+    logs:
+      receivers: [otlp]
+      processors: [..., unroll, ...]
+      exporters: [logging]
+```
+
+## Common unroll patterns
+
+The unroll processor only performs work if `log.body` is an iterable list—for
+example, a proper JSON array. But in real-world pipelines, log records aren't
+always so neatly structured. Sometimes, additional preprocessing is needed to
+convert raw log payloads into a format that the unroll processor can operate on.
+
+### Example: multiple JSON objects in a single log record
+
+Consider a case where multiple JSON objects are concatenated into a single log
+record, like this:
+
+```json
+{"@timestamp":"2025-09-19T02:20:17.920Z", "log.level": "INFO", "message":"initialized", "ecs.version": "1.2.0","service.name":"ES_ECS","event.dataset":"elasticsearch.server","process.thread.name":"main","log.logger":"org.elasticsearch.node.Node","elasticsearch.node.name":"es-test-3","elasticsearch.cluster.name":"elasticsearch"},{"type": "server", "timestamp": "2025-09-18T20:44:01,838-04:00", "level": "INFO", "component": "o.e.n.Node", "cluster.name": "elasticsearch", "node.name": "es-test", "message": "initialized" }
+```
+
+Here's how you can do that using the `transform` processor followed by `unroll`:
+
+```yaml
+receivers: ...
+
+processors:
+  transform:
+    error_mode: ignore
+    log_statements:
+      - context: log
+        statements:
+          - set(body, Split(body, "\"},"))
+  unroll: {}
+exporters: ...
+
+services:
+  pipelines:
+    logs:
+      receivers: [...]
+      processors: [transform, unroll]
+      exporters: [...]
+```
+
+This transform statement uses `Split` to separate the body into chunks using the
+`"},"` delimiter, producing a list-like body that the unroll processor can
+expand.
+
+## Summary
+
+This feature started from a simple need: make the Collector more versatile and
+capable of expanding log records.
+
+We tried the OTTL route, realized it wouldn't easily work, and upstreamed a
```

Comment:
Should OTTL be improved to allow safely adding new records? If no, why not? If yes, should we link to a tracking issue?

Reply:
Went ahead and added a bit to the summary to be a tad bit more explicit in the conclusion 😄